src/cpufreq.c

```diff
@@ -18,28 +18,125 @@
  *
  * Authors:
  *   Peter Holik <peter at holik.at>
+ *   Igor Solovyov <igor dot solovyov at gmail dot com>
  **/
 
+/**
+ * For quite a while, Linux kernel exposes the current CPU frequency via
+ * "/sys/devices/system/cpu/cpuX/cpufreq/scaling_cur_freq" file.
+ *
+ * Nevertheless, in CentOS 7.1 it was observed that kernel doesn't provide the file anymore, with
+ * the employed intel_pstate CPU frequency driver.
+ *
+ * The issue above has been fixed in https://github.com/torvalds/linux/commit/c034b02e213d271b98c45c4a7b54af8f69aaac1e .
+ * But up to now with for example, CentOS 7.2, the fix has not been back-ported yet.
+ *
+ * Note that there is another "cpuinfo_cur_freq" file under "/sys/devices/system/cpu/cpuX/cpufreq/" directory.
+ * This file exposes not exactly the same thing as "scaling_cur_freq" does, but close by nature.
+ * More details: http://www.pantz.org/software/cpufreq/usingcpufreqonlinux.html
+ *
+ * Overall it is better to have a workaround for monitoring some CPU frequency on even with kernels that lack
+ * the aforementioned fix.
+ *
+ * Also, an optional parameter "Path" is introduced to specify alternative CPU frequency path.
+ * Example:
+ * <Plugin cpufreq>
+ *     Path "/sys/devices/system/cpu/cpu%d/cpufreq/cpuinfo_cur_freq"
+ * </Plugin>
+ *
+ * Note, "%d" has to be used only once as CPU index placeholder.
+**/
+
 #include "collectd.h"
 #include "common.h"
 #include "plugin.h"
 
 #define MODULE_NAME "cpufreq"
+#define MAX_STR_L 256
 
 static int num_cpu = 0;
 
+static char const * freq_fname_def = "/sys/devices/system/cpu/cpu%d/cpufreq/"
+									 "scaling_cur_freq";
+static char freq_fname[MAX_STR_L] = {0};
+
+static const char *config_keys[] =
+{
+	"Path"
+};
+static int config_keys_num = STATIC_ARRAY_SIZE (config_keys);
+
+int check_format (char const * src)
+{
+	int pc = -2;
+	int d = -2;
+	char const * p;
+
+	for (p=src; *p; ++p) {
+		switch (*p) {
+			case '%':
+				if (-2 != pc) return 0;
+				pc=p-src;
+				break;
+
+			case 'd':
+				if (pc+1 == p-src) d=p-src;
+				break;
+		}
+	}
+
+	return pc+1==d;
+} /* check_format */
+
+static int cpufreq_config (char const * key, char const * value)
+{
+	if (strcasecmp ("Path", key) == 0) {
+		if (check_format (value)) {
+			sstrncpy (freq_fname, value, sizeof(freq_fname));
+		}
+		else {
+			WARNING ("cpufreq: Path parameter is wrong: %s", value);
+			return -1;
+		}
+	}
+
+	return 0;
```

Comment:
We prefer return (0). In general, please keep the same style as the existing code in this file.